### PR TITLE
MS teams conference initial commit

### PR DESCRIPTION
**Pull request by Cinojose**
Microsoft Teams Conference plugin.

Plugin to create Microsoft teams meetings during the incident creation. Plugin is tested as a part of the internal usage.
Screenshot 2023-06-07 at 12 32 21 PM

### CODE REVIEW

File: setup.py [hunk 1]
```diff
@@ -427,6 +427,7 @@ def run(self):
             "slack_contact = dispatch.plugins.dispatch_slack.plugin:SlackContactPlugin",
             "slack_conversation = dispatch.plugins.dispatch_slack.plugin:SlackConversationPlugin",
             "zoom_conference = dispatch.plugins.dispatch_zoom.plugin:ZoomConferencePlugin",
+            "teams_conference = dispatch.plugins.dispatch_microsoft_teams.conference.plugin:TeamsConferencePlugin",
```
Comment (mvilanova) [inline setup.py:430]: nit: should we prefix with the company name, like we do for other plugins?
```suggestion
            "microsoft_teams_conference = dispatch.plugins.dispatch_microsoft_teams.conference.plugin:MicrosoftTeamsConferencePlugin",
```

File: src/dispatch/plugins/dispatch_microsoft_teams/conference/config.py [hunk 2]
```diff
@@ -0,0 +1,28 @@
...
+    authority: str = Field(
+        title="MS team Authority URL",
+        description="Following format https://login.microsoftonline.com/Enter_the_Tenant_Id_Here.",
+    )
+    client_id: str = Field(
+        title="client id",
+        description="It is the Application (client) ID for the application you registered.",
+    )
+    secret: SecretStr = Field(
+        title="Azure Client Secret", description="This is the client secret created via Azure AD"
```
Comment (mvilanova) [inline src/dispatch/plugins/dispatch_microsoft_teams/conference/config.py:18]:
```suggestion
        title="Azure Client Secret", description="This is the client secret created via Azure AD."
```

File: src/dispatch/plugins/dispatch_microsoft_teams/conference/plugin.py [hunk 3]
```diff
@@ -0,0 +1,48 @@
...
+import logging
+from .client import MSTeamsClient
+from dispatch.decorators import apply, counter, timer
+
+
+logger = logging.getLogger(__name__)
+
+
+class TeamsConferencePlugin(ConferencePlugin):
+    title = "Microsoft teams Plugin - Conference Management"
```
Comment (mvilanova) [inline src/dispatch/plugins/dispatch_microsoft_teams/conference/plugin.py:16]:
```suggestion
    title = "Microsoft Teams Plugin - Conference Management"
```

File: src/dispatch/plugins/dispatch_microsoft_teams/conference/plugin.py [hunk 4]
```diff
@@ -0,0 +1,48 @@
...
+            )
+            meeting_info = client.create_meeting(name)
+
+            return {
+                "weblink": meeting_info["joinWebUrl"],
+                "id": meeting_info["id"],
+                "challenge": "",
+            }
+        except Exception as e:
+            logger.error(f"There is an error in creatig meeting {e}")
```
Comment (mvilanova) [inline src/dispatch/plugins/dispatch_microsoft_teams/conference/plugin.py:48]:
```suggestion
            logger.error(f"There was an error when attempting to create the meeting {e}")
```

File: src/dispatch/plugins/dispatch_microsoft_teams/conference/plugin.py [hunk 5]
```diff
@@ -0,0 +1,48 @@
...
+from .client import MSTeamsClient
+from dispatch.decorators import apply, counter, timer
+
+
+logger = logging.getLogger(__name__)
+
+
+class TeamsConferencePlugin(ConferencePlugin):
+    title = "Microsoft teams Plugin - Conference Management"
+    slug = "teams-conference"
```
Comment (mvilanova) [inline src/dispatch/plugins/dispatch_microsoft_teams/conference/plugin.py:17]:
```suggestion
    slug = "microsoft-teams-conference"
```

File: src/dispatch/plugins/dispatch_microsoft_teams/conference/plugin.py [hunk 6]
```diff
@@ -0,0 +1,48 @@
...
+from .config import TeamsConfiguration
+import logging
+from .client import MSTeamsClient
+from dispatch.decorators import apply, counter, timer
+
+
+logger = logging.getLogger(__name__)
+
+
+class TeamsConferencePlugin(ConferencePlugin):
```
Comment (mvilanova) [inline src/dispatch/plugins/dispatch_microsoft_teams/conference/plugin.py:15]:
```suggestion
class MicrosoftTeamsConferencePlugin(ConferencePlugin):
```

File: src/dispatch/plugins/dispatch_microsoft_teams/conference/config.py [hunk 7]
```diff
@@ -0,0 +1,28 @@
+from pydantic import Field, SecretStr
+
+from dispatch.config import BaseConfigurationModel
+
+
+class TeamsConfiguration(BaseConfigurationModel):
```
Comment (mvilanova) [inline src/dispatch/plugins/dispatch_microsoft_teams/conference/config.py:6]:
```suggestion
class MicrosoftTeamsConfiguration(BaseConfigurationModel):
```

**Review by mvilanova — CHANGES_REQUESTED** [review]
Looks good! Thanks for the PR

**Comment by Cinojose** [comment]
Fixed all the pr comments, please review. Thanks!

File: src/dispatch/plugins/dispatch_microsoft_teams/conference/client.py [hunk 8]
```diff
@@ -0,0 +1,75 @@
...
+            graph_data = requests.post(  # Use token to call downstream service
+                url=f"https://graph.microsoft.com/v1.0/users/{self.user_id}/onlineMeetings",
+                headers={"Authorization": "Bearer " + result["access_token"]},
+                json=data,
+            ).json()
+            logger.info("Graph API call result: ")
+            logger.info(json.dumps(graph_data, indent=2))
+            return graph_data
+        else:
+            return dict()
```
Comment (mvilanova) [inline src/dispatch/plugins/dispatch_microsoft_teams/conference/client.py:75]: nit
```suggestion

        return dict()
```

**Review by mvilanova — APPROVED** [review]

**Comment by Cinojose** [comment]
> @Cinojose can you fix the linting errors? https://github.com/Netflix/dispatch/actions/runs/5262948502/jobs/9528851603?pr=3474

Done, please check.

**Pull request merged**